File: src/matrix_ops.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Matrix = @import("matrix.zig").Matrix;
const MatrixError = @import("matrix.zig").MatrixError;
const backend = @import("backend.zig");

extern fn matrix_add(context: *anyopaque, a: [*]const f32, b: [*]const f32, c: [*]const f32, n: c_uint) void;
extern fn matrix_sub(context: *anyopaque, a: [*]const f32, b: [*]const f32, c: [*]const f32, n: c_uint) void;
extern fn matrix_mul(context: *anyopaque, a: [*]const f32, b: [*]const f32, c: [*]const f32, a_rows: c_uint, a_cols: c_uint, b_rows: c_uint, b_cols: c_uint) void;
extern fn matrix_scale(context: *anyopaque, scale: f32, a: [*]const f32, b: [*]const f32, n: c_uint) void;

pub fn Add(allocator: Allocator, a: Matrix(f32), b: Matrix(f32)) MatrixError!Matrix(f32) {
    if (a.rows != b.rows or a.cols != b.cols) {
        return MatrixError.MissMatchShape;
    }

    const result = Matrix(f32).init(allocator, a.rows, b.cols) catch {
        return MatrixError.FailAlloc;
    };

    const ctx = backend.GetInstance();
    matrix_add(ctx.context, a.values.ptr, b.values.ptr, result.values.ptr, a.rows * a.cols);

    return result;
}

pub fn Subtract(allocator: Allocator, a: Matrix(f32), b: Matrix(f32)) MatrixError!Matrix(f32) {
    if (a.rows != b.rows or a.cols != b.cols) {
        return MatrixError.MissMatchShape;
    }

    const result = Matrix(f32).init(allocator, a.rows, a.cols) catch {
        return MatrixError.FailAlloc;
    };

    const ctx = backend.GetInstance();
    matrix_sub(ctx.context, a.values.ptr, b.values.ptr, result.values.ptr, a.rows * a.cols);

    return result;
}

pub fn Multiply(allocator: Allocator, a: Matrix(f32), b: Matrix(f32)) MatrixError!Matrix(f32) {
    if (a.cols != b.rows) {
        return MatrixError.MissMatchShape;
    }

    const result = Matrix(f32).init(allocator, a.rows, b.cols) catch {
        return MatrixError.FailAlloc;
    };

    const ctx = backend.GetInstance();
    matrix_mul(ctx.context, a.values.ptr, b.values.ptr, result.values.ptr, a.rows, a.cols, b.rows, b.cols);

    return result;
}

pub fn Scale(allocator: Allocator, a: Matrix(f32), scale: f32) MatrixError!Matrix(f32) {
    const result = Matrix(f32).init(allocator, a.rows, a.cols) catch {
        return MatrixError.FailAlloc;
    };

    const ctx = backend.GetInstance();
    matrix_scale(ctx.context, scale, a.values.ptr, result.values.ptr, a.rows * a.cols);

    return result;
}

File: backends/cpu/matrix_ops.c
#include "backend.h"
#include <math.h>

void matrix_add(void* context, float* a, const float* b, unsigned int n) {
    for (int i = 0; i < n; ++i) {
        a[i] += b[i];
    }
}

void matrix_sub(void* context, float* a, const float* b, unsigned int n) {
    for (int i = 0; i < n; ++i) {
        a[i] -= b[i];
    }
}

void matrix_mul(void* context, const float* a, const float* b, float* c, unsigned int a_rows, unsigned int a_cols, unsigned int b_rows, unsigned int b_cols) {
    for (int i = 0; i < a_rows; ++i) {
        for (int j = 0; j < b_cols; ++j) {
            float sum = 0;
            for (int k = 0; k < a_cols; ++k) {
                sum += a[a_cols * i + k] * b[b_cols * k + j];
            }
            c[b_cols * i + j] = sum;
        }
    }
}

void matrix_scale(void* context, float scale, float* a, unsigned int n) {
    for (int i = 0; i < n; ++i) {
        a[i] *= scale;
    }
}

void f_sigmoid(void* context, float* a, unsigned int n) {
    for (int i = 0; i < n; i++) {
        a[i] = 1.0 / (1.0 + expf(-a[i]));
    }
}

void f_relu(void* context, float* a, unsigned int n) {
    for (int i = 0; i < n; i++) {
        a[i] = (a[i] + fabsf(a[i])) / 2;
    }
}

void f_softmax(void* context, float* a, unsigned int n) {
    float sum = 0;

    for (int i = 0; i < n; i++) {
        a[i] = expf(a[i]);
        sum +=  a[i];
    }

    for (int i = 0; i < n; i++) {
        a[i] /= sum;
    }
}

void f_tanh(void* context, float* a, unsigned int n) {
    for (int i = 0; i < n; i++) {
        float ei = expf(a[i]);
        float nei = expf(-a[i]);
        a[i] = (ei - nei) / (ei + nei);
    }
}